### PR TITLE
[RF] Improve recovery from invalid function values

### PR DESCRIPTION
**Pull request by hageboeck**
If the parameters of a function are outside of the definition range,
all kinds of errors might occur. PDFs might be negative, have negative
integrals, coefficients for the summation of PDFs might degenerate, ...

Previously, RooFit was just returning a high function value to Minuit.
This value was always the same, though.
Like this, Minuit cannot compute a gradient to get out of this region,
because all bad parameter points yield the same function value.

With this commit, RooFit can pack information about the "badness" of a
parameter point into the mantissa of a NaN, e.g. how much negative the
values of a PDF came out or how far the sum of PDF coefficients is from
1. This information is packed into NaN using "RooNaNPacker", and passed
through the computation graph. It is finally unpacked in RooMinimizerFcn
before being handed to the minimiser. This allows for the calculation of
gradients, enabling the fitter to recover from an invalid state more
easily.

When NaNs are being unpacked, they are presented to Minuit as maximum
function value + badness * penaltyStrength, so a gradient can be
computed that points away from the bad region. The maximum function
value comes from previous minimisation steps, the badness is the payload
of the packed NaN, and the strength can be set from the outside.

If the minimisation starts out in an undefined region, and the maximum
function value is not (yet) known, 0. + badness * strength is used.
Once a well-defined region is found, the valid function values are
offset such that they are all lower than 0. In this way, Minuit is
encouraged to step into the well-defined regions, and sees a rapidly
rising function when stepping out. When fit results are saved, though,
this offset is subtracted in order to recover "correct" NLL values.

Concretely:
- Add RooNaNPacker.
- Add a function to set the magnitude of the penalty term that's
supposed to drive the minimiser away from undefined regions.
- Add a function to RooMinimizerFcn that returns a possible artificial
offset. This helps to hide any offsetting from the outside world.
- When saving a fit result, remove any offset, and copy the number of
invalid function values into the fit result (was broken previously).

### CODE REVIEW

File: roofit/roofit/src/RooPoisson.cxx [hunk 1]
```diff
@@ -65,8 +66,11 @@ RooPoisson::RooPoisson(const char *name, const char *title,
 Double_t RooPoisson::evaluate() const
 {
   Double_t k = _noRounding ? x : floor(x);
-  if(_protectNegative && mean<0)
-    return 1e-3;
+  if(_protectNegative && mean<0) {
```
Comment (lmoneta) [inline roofit/roofit/src/RooPoisson.cxx:69]: is _protecNegative set by default i RooPoisson ?

**Review by lmoneta — APPROVED** [review]

**Comment by lmoneta** [comment]
Very nice contribution !
Should we add maybe also a tutorial similar to the degenerate coefficient test ?

File: roofit/roofit/src/RooPoisson.cxx (hunk 1 above)
Comment (hageboeck) [reply roofit/roofit/src/RooPoisson.cxx:69]: It's actually off by default. I will enable it in an extra commit.

**Comment by hageboeck** [comment]
> Should we add maybe also a tutorial similar to the degenerate coefficient test ?

I will convert the test into one.

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
@phsft-bot build

**Pull request merged**